### PR TITLE
Postgres Write 3/4: Schema validation and transformation for creating new annotations in Postgres

### PR DESCRIPTION
**Pull request by seanh**
This pull request adds:

1. Transformation of annotation data and document metadata from the client into the format that the `storage` module will need it in (see <https://github.com/hypothesis/h/pull/3153>)

   The format of the "document" sub-object that the client POSTs in annotation create and update requests is quite far from the format that we need in order to create/update the `Document`, `DocumentMeta` and `DocumentURI` objects in the database. The new `h.api.parse_document_claims` module handles transforming this dict into two lists that are in a more convenient format for `storage` to later use to create the db objects. The code in here is adapted from existing code in `migrate.py` and `elastic.py`.

   `schemas` calls `parse_document_claims` to do this, so when the data comes out of schema validation it's in the right/convenient format.

2. Additional validation of the data from the client so that `storage` and the db don't crash.

   Now that we have a real database we can't just throw any crap into it, either the db or our own code will crash if required fields are missing or if known fields are the wrong type. Minimal additional JSON Schema validation has been added (only to the create annotation API and only when postgres_write is on) to prevent these crashes and send proper validation errors back.

`schemas` only does "structural" validation and transformations, i.e. work that doesn't require looking in the db. Some further "semantic" validation and transformation that does require reading from the will be done in `storage`.

### CODE REVIEW

**Comment by seanh** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;0xedAaiS&#x2F;282-postgres-write-annotation-creates-and-updates-to-postgres-when-the-postgres-write-feature-flag-is-on"> Postgres: write annotation creates and updates to postgres when the postgres_write feature flag is on</a>

File: h/api/schemas.py [hunk 1]
```diff
@@ -90,9 +205,75 @@ class CreateAnnotationSchema(object):
 
     def __init__(self, request):
         self.request = request
-        self.structure = AnnotationSchema()
+        if self.request.feature('postgres_write'):
+            self.structure = AnnotationSchema()
+        else:
+            self.structure = LegacyAnnotationSchema()
 
     def validate(self, data):
```
Comment (chdorner) [inline h/api/schemas.py]: Please help me understand something here.
By moving the previous implementation of `validate` into `_legacy_validate` and depending on the `postgres_write` feature flag either return the dict that has been sanitized by `validate` (for flag on) or `_legacy_validate` (for flag off), doesn't that mean that we start storing things like `target_uri` instead of the `uri` for the Elasticsearch annotations as soon as we turn the feature flag on?
Since turning the feature flag on means we write to both Elasticsearch and to Postgres. Which means that the JSON structure we return in the API will render `target_uri` instead of `uri` which is what we need.
Or am I missing something here?

File: h/api/schemas.py [hunk 2]
```diff
@@ -59,6 +61,119 @@ class AnnotationSchema(JSONSchema):
...
+            },
+            'text': {
+                'type': 'string',
+            },
+            'uri': {
+                'type': 'string',
+            },
+        },
+        'required': [
+            'permissions',
```
Comment (chdorner) [inline h/api/schemas.py]: We might want to add the `uri` to the required fields as well, since currently the column does not allow `NULL` values and it doesn't provide a default value.

File: h/api/schemas.py [hunk 3]
```diff
@@ -90,9 +205,75 @@ class CreateAnnotationSchema(object):
...
+        # annotation.target_selectors whose value is the first selector in
+        # the client'ss target.selectors list.
+        # Anything else in the target dict, and any selectors after the first,
+        # are discarded.
+        target = appstruct.pop('target', [])
+        if target:  # Replies and page notes don't have 'target'.
+            target = target[0]  # Multiple targets are ignored.
+            new_appstruct['target_selectors'] = target['selector']
+
+        new_appstruct['groupid'] = appstruct.pop('group', '__world__')
```
Comment (chdorner) [inline h/api/schemas.py]: We already have `__world__` set as the default in the database, I like it better here, maybe we should remove the default value from the database? so we don't have to maintain the same value in two places?

File: h/api/schemas.py [hunk 4]
```diff
@@ -90,9 +205,75 @@ class CreateAnnotationSchema(object):
...
+        # groupid is added to the reply annotation later by the storage code.
+        # Here we just delete any group sent by the client from replies.
+        if new_appstruct['references'] and 'groupid' in new_appstruct:
+            del new_appstruct['groupid']
+
+        new_appstruct['extra'] = appstruct
+
+        # Transform the "document" dict that the client posts into a convenient
+        # format for creating DocumentURI and DocumentMeta objects later.
+        document_data = appstruct.pop('document', {})
```
Comment (chdorner) [inline h/api/schemas.py]: Would it make sense to pass separate copies of `document_data`into the next two function calls? So we don't rely on these two functions not mutating this dictionary.

File: h/api/parse_document_claims.py [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+                                path_prefix=keypath)
+            else:
+                if not isinstance(value, list):
+                    value = [value]
+
+                document_meta_dicts.append({
+                    'type': '.'.join(keypath),
+                    'value': value,
+                    'claimant': claimant,
+                    'claimant_normalized':
```
Comment (chdorner) [inline h/api/parse_document_claims.py]: There's no need to set the `claimant_normalized` manually, the `claimant` setter takes care of this.

File: h/api/parse_document_claims.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+        if link.keys() == ['href'] and link['href'].startswith('doi:'):
+            continue
+
+        uri_ = link['href']
+        type_ = None
+
+        # Handle Highwire PDF links.
+        if set(link.keys()) == set(['href', 'type']):
+            if link['type'] == 'application/pdf':
+                type_ = 'highwire-pdf'
```
Comment (chdorner) [inline h/api/parse_document_claims.py]: I was thinking about this case recently. Instead of relying on the annotator doing the right thing ([this](https://github.com/openannotation/annotator/blob/v1.2.x/src/plugin/document.coffee#L122-L128)), wouldn't it be safer to get the `highwire-pdf` DocumentURI from the highwire metadata, similar to what we do for the doi uris?

File: h/api/schemas.py (hunk 1 above)
Comment (seanh) [inline h/api/schemas.py]: Oh damn, I forgot that the feature flag was supposed to mean write to both storage backends together, this code is intended (when combined with the next pull request) to write to postgres only when the feature flag is on.

Hopefully it won't be too difficult to fix it to write different data to each instead though..

Comment (seanh) [inline h/api/schemas.py]: This is fixed now (in this pr at least)

File: h/api/schemas.py (hunk 2 above)
Comment (seanh) [inline h/api/schemas.py]: This one is actually okay because the `validate()` method does `new_appstruct['target_uri'] = appstruct.pop('uri', '')`.

In the future we probably should add uri to the required fields I agree, because why would we want people posting annotations with no uri? But for this pr, I just wanted to put in the absolute minimum validation to stop users from being able to crash it.

File: h/api/schemas.py (hunk 3 above)
Comment (seanh) [inline h/api/schemas.py]: Schema validation and transformation does seem like the correct place to insert default values, so I agree. Can I do that in a separate PR though? I would like to get postgres PRs 1-4 all merged so that the whole thing can be tested together, before making a lot more changes to it.

File: h/api/schemas.py (hunk 4 above)
Comment (seanh) [inline h/api/schemas.py]: Sure, seems no harm in being defensive

File: h/api/parse_document_claims.py (hunk 5 above)
Comment (seanh) [inline h/api/parse_document_claims.py]: Done

File: h/api/parse_document_claims.py (hunk 6 above)
Comment (seanh) [inline h/api/parse_document_claims.py]: So you mean we'd just add read the `data["highwire"]["pdf_url"]` list directly, and add a `DocumentURI` for each PDF URL in that list? Can do..

Comment (seanh) [inline h/api/parse_document_claims.py]: Done

File: h/api/parse_document_claims.py (hunk 5 above)
Comment (seanh) [inline h/api/parse_document_claims.py]: Done

**Comment by seanh** [comment]
Ok, I'll see if I can tidy up the history a bit...

File: h/api/schemas.py (hunk 3 above)
Comment (seanh) [inline h/api/schemas.py]: Created an issue to do this separately https://github.com/hypothesis/h/issues/3177

**Comment by seanh** [comment]
@chdorner I think I've responded to all the comments now, and I've completely rewritten the history of this one so it's nicer

File: h/api/schemas.py (hunk 2 above)
Comment (chdorner) [inline h/api/schemas.py]: Sounds good to me

File: h/api/parse_document_claims.py [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+    """
+    Return PDF document URI dicts for the given 'highwire' document metadata.
+
+    Process a document.highwire dict that the client submitted as part of an
+    annotation create or update request and return document URI dicts for all
+    of the PDF document equivalence claims that it makes.
+
+    """
+    document_uris = []
+    hwpdfvalues = highwire_dict.get('pdf', [])
```
Comment (chdorner) [inline h/api/parse_document_claims.py]: I think this should be `highwire_dict.get('pdf_url', [])`, as that's what the [annotator does](https://github.com/openannotation/annotator/blob/v1.2.x/src/plugin/document.coffee#L122-L128).
Or did you find some data that had the PDF url with the key `pdf`?

File: h/api/parse_document_claims.py [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+        if link.keys() == ['href'] and link['href'] == claimant:
+            continue
+
+        # Disregard doi links as these are being added separately from the
+        # highwire and dc metadata later on.
+        if link.keys() == ['href'] and link['href'].startswith('doi:'):
+            continue
+
+        # Disregard Highwire PDF links as these are being added separately from
+        # the highwire and dc metadata later on.
```
Comment (chdorner) [inline h/api/parse_document_claims.py]: Nit, the comment should not say `and dc`.

**Comment by chdorner** [comment]
Looking great! Only thing need to resolve is how it reads the `highwire-pdf` url (unless I'm wrong about my assumption). Then it's good to merge

File: h/api/parse_document_claims.py (hunk 7 above)
Comment (seanh) [inline h/api/parse_document_claims.py]: Good catch, fixed

File: h/api/parse_document_claims.py (hunk 8 above)
Comment (seanh) [inline h/api/parse_document_claims.py]: Done
